src/rust/engine/graph/src/lib.rs

```diff
@@ -14,28 +14,111 @@ use std::collections::{HashMap, HashSet, VecDeque};
 
 use futures::future::{self, Future};
-use petgraph::stable_graph::{StableDiGraph, StableGraph};
+use futures::sync::oneshot;
+use petgraph::graph::DiGraph;
 use petgraph::Direction;
 
 use boxfuture::{BoxFuture, Boxable};
 pub use node::{EntryId, Node, NodeContext, NodeError, NodeTracer, NodeVisualizer};
 
 type FNV = BuildHasherDefault<FnvHasher>;
 
-type PGraph<N> = StableDiGraph<Entry<N>, (), u32>;
+type PGraph<N> = DiGraph<Entry<N>, (), u32>;
 
-type EntryStateField<Item, Error> = future::Shared<BoxFuture<Item, Error>>;
+///
+/// A token that uniquely identifies one run of a Node in the Graph. Each run of a Node (via
+/// `N::Context::spawn`) has a different RunToken associated with it. When a run completes, if
+/// the current RunToken of its Node no longer matches the RunToken of the spawned work (because
+/// the Node was `cleared`), the work is discarded. See `Entry::complete` for more information.
+///
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct RunToken(u32);
+
+impl RunToken {
+  fn initial() -> RunToken {
+    RunToken(0)
+  }
+
+  fn next(&self) -> RunToken {
+    RunToken(self.0 + 1)
+  }
+}
+
+///
+/// A token associated with a Node that is incremented whenever its output value has (or might
+/// have) changed. When a dependent consumes a dependency at a particular generation, that
+/// generation is recorded on the consuming edge, and can later used to determine whether the
+/// inputs to a node have changed.
+///
+/// Unlike the RunToken (which is incremented whenever a node re-runs), the Generation is only
+/// incremented when the output of a node has changed.
+///
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct Generation(u32);
+
+impl Generation {
+  fn initial() -> Generation {
+    Generation(0)
+  }
+
+  fn next(&self) -> Generation {
+    Generation(self.0 + 1)
+  }
+}
 
-struct EntryState<N: Node> {
-  field: EntryStateField<N::Item, N::Error>,
-  start_time: Instant,
+enum EntryState<N: Node> {
+  // A node that has either been explicitly cleared, or has not yet started Running for the first
+  // time. In this state there is no need for a dirty bit because the generation is either in its
+  // initial state, or has been explicitly incremented when the node was cleared.
+  //
+  // The previous_result value is _not_ a valid value for this Entry: rather, it is preserved in
+  // order to compute the generation value for this Node by comparing it to the new result the next
+  // time the Node runs.
+  NotStarted {
+    run_token: RunToken,
+    generation: Generation,
+    previous_result: Option<Result<N::Item, N::Error>>,
+  },
+  // A node that is running. A running node that has been marked dirty re-runs rather than
+  // completing.
+  //
+  // The `previous_result` value for a Running node is not a valid value. See NotStarted.
+  Running {
+    run_token: RunToken,
+    generation: Generation,
+    start_time: Instant,
+    waiters: Vec<oneshot::Sender<Result<(N::Item, Generation), N::Error>>>,
```

Comment:
Does this signature suggest that `Error`s are always a new generation, even if they're the same? Possibly this should be a `oneshot::Sender<(Result<N::Item, N::Error>, Generation)>` so that identical errors don't trigger downstream invalidations?